File: app/src/lib/scripts.ts
/**
 * Scripts Layer: Types and utilities for YAML scenario handling
 *
 * This module defines the core types for the English-Liminal script system
 * and provides functions for loading, validating, and managing scenarios.
 */

// ============================================================================
// Core Step Types
// ============================================================================

/**
 * Listen step: User listens to audio/text
 */
export type ListenStep = {
  type: 'listen';
  prompt: string;
  content: string;
  target?: string;
};

/**
 * Speak Check step: User repeats a phrase, gets pronunciation feedback
 */
export type SpeakCheckStep = {
  type: 'speak_check';
  prompt: string;
  content: string;
  ref_text: string;
  eval: 'pronunciation' | 'fluency' | 'both';
  hints?: string[];
};

/**
 * Contrast step: Multiple choice or comparison exercise
 */
export type ContrastStep = {
  type: 'contrast';
  prompt: string;
  question: string;
  answers: {
    text: string;
    correct: boolean;
    explanation: string;
  }[];
};

/**
 * Apply to Life step: User creates their own example
 */
export type ApplyToLifeStep = {
  type: 'apply_to_life';
  prompt: string;
  instruction: string;
  examples?: string[];
  recording: boolean;
};

/**
 * Union type for all step types
 */
export type Step =
  | ListenStep
  | SpeakCheckStep
  | ContrastStep
  | ApplyToLifeStep;

// ============================================================================
// Context and Metadata Types
// ============================================================================

/**
 * Context triggers: when and how to show this scenario
 */
export type ContextTrigger = {
  time?: 'morning' | 'afternoon' | 'evening' | 'night';
  location?: 'home' | 'work' | 'commute' | 'cafe' | 'gym';
  duration_max_sec?: number;
  user_level?: 'beginner' | 'intermediate' | 'advanced';
};

/**
 * Rehearsal configuration: how the scenario should repeat over time
 */
export type RehearsalConfig = {
  decay_alpha: number;           // 0.7-0.9: speed of memory decay
  next_ping_sec_min: number;     // Minimum seconds until next ping
  next_ping_sec_max: number;     // Maximum seconds until next ping
  importance?: 'low' | 'medium' | 'high';
  tags?: string[];
};

// ============================================================================
// Main Script Type
// ============================================================================

/**
 * Complete script/scenario definition
 */
export type Script = {
  id: string;
  title: string;
  description?: string;
  context_triggers?: ContextTrigger[];
  goals?: string[];
  steps: Step[];
  rehearsal?: RehearsalConfig;
};

/**
 * Script metadata (for listing scenarios without loading full content)
 */
export type ScriptMeta = {
  id: string;
  title: string;
  description?: string;
  duration_sec?: number;
  tags?: string[];
  difficulty?: 'beginner' | 'intermediate' | 'advanced';
};

// ============================================================================
// Script Loading and Management
// ============================================================================

/**
 * In-memory cache for loaded scripts
 */
const scriptCache = new Map<string, Script>();

/**
 * Load a script by ID from YAML files
 *
 * @param id - Script ID (e.g., 'morning-warmup-01')
 * @returns Parsed and validated Script object
 * @throws Error if script not found or invalid
 *
 * @example
 * const script = await loadScript('morning-warmup-01');
 * console.log(script.title); // "Morning Warmup: A/An & H-dropping"
 */
export async function loadScript(id: string): Promise<Script> {
  // Check cache first
  if (scriptCache.has(id)) {
    return scriptCache.get(id)!;
  }

  // TODO: Implementation in Issue #3
  // 1. Load YAML file from app/src/data/${id}.yaml
  // 2. Parse with js-yaml
  // 3. Validate with zod
  // 4. Cache result
  // 5. Return Script object

  throw new Error(`Script loading not yet implemented. Run Issue #3 first.`);
}

/**
 * Get list of all available scripts
 *
 * @returns Array of script metadata
 *
 * @example
 * const scripts = await listAvailableScripts();
 * scripts.forEach(s => console.log(s.title));
 */
export async function listAvailableScripts(): Promise<ScriptMeta[]> {
  // TODO: Implementation in Issue #3
  // 1. Scan app/src/data/ directory
  // 2. Load metadata from each YAML
  // 3. Return array of ScriptMeta

  return [];
}

/**
 * Preload a script into cache
 *
 * @param id - Script ID to preload
 */
export async function preloadScript(id: string): Promise<void> {
  await loadScript(id);
}

/**
 * Clear script cache (useful for testing or reload)
 */
export function clearScriptCache(): void {
  scriptCache.clear();
}

// ============================================================================
// Validation Schemas (for zod)
// ============================================================================

/**
 * Zod schemas will be implemented in Issue #3
 * Example structure:
 *
 * import { z } from 'zod';
 *
 * const StepSchema = z.union([
 *   z.object({ type: z.literal('listen'), ... }),
 *   z.object({ type: z.literal('speak_check'), ... }),
 *   ...
 * ]);
 *
 * const ScriptSchema = z.object({
 *   id: z.string(),
 *   title: z.string(),
 *   steps: z.array(StepSchema),
 *   ...
 * });
 */

// ============================================================================
// Exports
// ============================================================================

export default {
  loadScript,
  listAvailableScripts,
  preloadScript,
  clearScriptCache,
};

File: core/src/scripts.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StepType {
    Listen,
    SpeakCheck,
    Contrast,
    ApplyToLife,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Answer {
    Simple(String),
    Rich {
        text: String,
        #[serde(default)]
        correct: bool,
        #[serde(default)]
        explanation: Option<String>,
    },
}

impl Answer {
    pub fn text(&self) -> &str {
        match self {
            Answer::Simple(s) => s,
            Answer::Rich { text, .. } => text,
        }
    }

    pub fn is_correct(&self) -> bool {
        match self {
            Answer::Simple(_) => false,
            Answer::Rich { correct, .. } => *correct,
        }
    }

    pub fn explanation(&self) -> Option<&str> {
        match self {
            Answer::Simple(_) => None,
            Answer::Rich { explanation, .. } => explanation.as_deref(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Step {
    pub r#type: StepType,
    pub prompt: String,
    #[serde(default)]
    pub content: Option<String>,
    #[serde(default)]
    pub ref_text: Option<String>,
    #[serde(default)]
    pub question: Option<String>,
    #[serde(default)]
    pub answers: Option<Vec<Answer>>,
    #[serde(default)]
    pub hints: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Rehearsal {
    #[serde(default = "default_decay")]
    pub decay_alpha: f32,           // 0.7..0.9
    #[serde(default = "default_ping_min")]
    pub next_ping_sec_min: u32,     // 90
    #[serde(default = "default_ping_max")]
    pub next_ping_sec_max: u32,     // 3600
}

fn default_decay() -> f32 { 0.82 }
fn default_ping_min() -> u32 { 90 }
fn default_ping_max() -> u32 { 3600 }

impl Default for Rehearsal {
    fn default() -> Self {
        Self {
            decay_alpha: default_decay(),
            next_ping_sec_min: default_ping_min(),
            next_ping_sec_max: default_ping_max(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Script {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub context_triggers: serde_json::Value,
    #[serde(default)]
    pub goals: Vec<String>,
    pub steps: Vec<Step>,
    #[serde(default)]
    pub rehearsal: Rehearsal,
}
